correctly transfer username validation to auth0

etwillbefine opened the pull request:
<!--- 

**IMPORTANT:** Please submit pull requests to [alexkappa/terraform-provider-auth0](https://github.com/alexkappa/terraform-provider-auth0). This helps maintainers organize work more efficiently.

See what makes a good Pull Request at : https://github.com/alexkappa/terraform-provider-auth0/blob/master/.github/CONTRIBUTING.md#pull-requests 

--->
### Proposed Changes

* Correctly sending username validation to Auth0

Fixes #238

#### Acceptance Test Output

```
make testacc TESTS=TestAccConnection
==> Checking that code complies with gofmt requirements...
?   	github.com/terraform-providers/terraform-provider-auth0	[no test files]
=== RUN   TestAccConnection
--- PASS: TestAccConnection (2.01s)
=== RUN   TestAccConnectionAD
--- PASS: TestAccConnectionAD (0.85s)
=== RUN   TestAccConnectionAzureAD
--- PASS: TestAccConnectionAzureAD (7.16s)
=== RUN   TestAccConnectionOIDC
--- PASS: TestAccConnectionOIDC (1.80s)
=== RUN   TestAccConnectionWithEnbledClients
--- PASS: TestAccConnectionWithEnbledClients (19.63s)
=== RUN   TestAccConnectionSMS
--- PASS: TestAccConnectionSMS (0.90s)
=== RUN   TestAccConnectionEmail
--- PASS: TestAccConnectionEmail (6.55s)
=== RUN   TestAccConnectionSalesforce
--- PASS: TestAccConnectionSalesforce (0.90s)
=== RUN   TestAccConnectionGoogleOAuth2
--- PASS: TestAccConnectionGoogleOAuth2 (0.83s)
=== RUN   TestAccConnectionFacebook
--- PASS: TestAccConnectionFacebook (7.67s)
=== RUN   TestAccConnectionApple
--- PASS: TestAccConnectionApple (1.73s)
=== RUN   TestAccConnectionLinkedin
--- PASS: TestAccConnectionLinkedin (7.84s)
=== RUN   TestAccConnectionGitHub
--- PASS: TestAccConnectionGitHub (0.97s)
=== RUN   TestAccConnectionConfiguration
--- PASS: TestAccConnectionConfiguration (7.83s)
=== RUN   TestAccConnectionSAML
--- PASS: TestAccConnectionSAML (7.75s)
PASS
coverage: 27.3% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0	74.726s	coverage: 27.3% of statements
?   	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/debug	[no test files]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/random	0.262s	coverage: 0.0% of statements [no tests to run]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/validation	0.288s	coverage: 0.0% of statements [no tests to run]
?   	github.com/terraform-providers/terraform-provider-auth0/version	[no test files]
...
```

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

#### Note: Contains breaking changes:
- Moved `options.requires_username` to `options.validation` 
- Changed Data Structure of `options.validation`

etwillbefine commented:
As suggested in the README `make test` runs successful. But for the CI run (`make testacc ..`)  it fails locally and on Travis, missing Provider configuration. Help appreciated.

alexkappa commented:
Hi, thanks for giving this a go!

Regarding the tests failing on CI, its kind of expected as PRs wont have access to secrets. You should run them locally though and paste the output in the right section in the description.

As this is a breaking change, we should add a migration so as to not break existing users.

I wasn't able to see any changes to flattening/expanding. Do the right values get passed along to the sdk? I’m on mobile so I might have missed something.

Cheers,
Alex

etwillbefine commented:
The Auth0 Client supports [`validation`](https://github.com/go-auth0/auth0/blob/master/management/connection.go#L157) already with a generic interface type. 

During Testing I realised that the `requires_username` actually works. Not sure how one of my connection had it set to false and true in Terraform but when testing with a fresh Connection it works. When sending requires_username within the validation object Auth0 returns a Bad Request that this property is not allowed. That means Auth0's Documentation is not correct about requires_username property. https://auth0.com/docs/connections/references/options-mgmt-api

I created the corresponding mapping for validation, removed changes related to `requires_username` and added Acceptance Test Output to the original Description.

etwillbefine commented on auth0/resource_auth0_connection.go:
```diff
@@ -181,9 +204,8 @@ var connectionSchema = map[string]*schema.Schema{
-					Type:        schema.TypeBool,
-					Optional:    true,
-					Description: "Indicates whether or not the user is required to provide a username in addition to an email address",
+					Type:     schema.TypeBool,
+					Optional: true,
```
added description again

etwillbefine commented on auth0/resource_auth0_connection_test.go:
```diff
@@ -110,11 +112,17 @@ resource "auth0_connection" "my_connection" {
 		brute_force_protection = true
 		import_mode = false
 		disable_signup = false
-		requires_username = true
```
cant really revert as in same commit. moved the property to its previous position

etwillbefine commented on auth0/resource_auth0_connection.go:
```diff
@@ -81,8 +81,31 @@ var connectionSchema = map[string]*schema.Schema{
 				"validation": {
-					Type:     schema.TypeMap,
-					Elem:     &schema.Schema{Type: schema.TypeString},
+					Type:     schema.TypeList,
```
Thats going to take same time as Im new to Go and Terraform Provider Migrations. Will try to provide it within the next few days.

alexkappa replied:
I understand :) I’m here if you need any help.

For inspiration, you can see the migration of connections schema a while back.

etwillbefine commented:
I added the migration to move `validation.min|max` to `validation.username.min|max`. I'm not sure if I need to make sure to convert the previous string value to int. `structure_auth0_validation.go` contains already converting to Int value. 

Here is an updated Test Output:
```
make testacc TESTS=TestAccConnection
==> Checking that code complies with gofmt requirements...
?   	github.com/terraform-providers/terraform-provider-auth0	[no test files]
=== RUN   TestAccConnection
--- PASS: TestAccConnection (2.35s)
=== RUN   TestAccConnectionAD
--- PASS: TestAccConnectionAD (0.89s)
=== RUN   TestAccConnectionAzureAD
--- PASS: TestAccConnectionAzureAD (7.19s)
=== RUN   TestAccConnectionOIDC
--- PASS: TestAccConnectionOIDC (1.97s)
=== RUN   TestAccConnectionWithEnbledClients
--- PASS: TestAccConnectionWithEnbledClients (13.65s)
=== RUN   TestAccConnectionSMS
--- PASS: TestAccConnectionSMS (7.05s)
=== RUN   TestAccConnectionEmail
--- PASS: TestAccConnectionEmail (1.66s)
=== RUN   TestAccConnectionSalesforce
--- PASS: TestAccConnectionSalesforce (7.01s)
=== RUN   TestAccConnectionGoogleOAuth2
--- PASS: TestAccConnectionGoogleOAuth2 (0.93s)
=== RUN   TestAccConnectionFacebook
--- PASS: TestAccConnectionFacebook (7.74s)
=== RUN   TestAccConnectionApple
--- PASS: TestAccConnectionApple (1.64s)
=== RUN   TestAccConnectionLinkedin
--- PASS: TestAccConnectionLinkedin (7.90s)
=== RUN   TestAccConnectionGitHub
--- PASS: TestAccConnectionGitHub (0.97s)
=== RUN   TestAccConnectionConfiguration
--- PASS: TestAccConnectionConfiguration (7.73s)
=== RUN   TestAccConnectionSAML
--- PASS: TestAccConnectionSAML (1.69s)
PASS
coverage: 26.9% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0	70.714s	coverage: 26.9% of statements
?   	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/debug	[no test files]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/random	0.246s	coverage: 0.0% of statements [no tests to run]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/validation	0.146s	coverage: 0.0% of statements [no tests to run]
?   	github.com/terraform-providers/terraform-provider-auth0/version	[no test files]
```

etwillbefine commented:
Here is an updated Test Output:

make testacc TESTS=TestAccConnection
==> Checking that code complies with gofmt requirements...
?   	github.com/terraform-providers/terraform-provider-auth0	[no test files]
=== RUN   TestAccConnection
--- PASS: TestAccConnection (2.20s)
=== RUN   TestAccConnectionAD
--- PASS: TestAccConnectionAD (0.89s)
=== RUN   TestAccConnectionAzureAD
--- PASS: TestAccConnectionAzureAD (7.25s)
=== RUN   TestAccConnectionOIDC
--- PASS: TestAccConnectionOIDC (1.68s)
=== RUN   TestAccConnectionWithEnbledClients
--- PASS: TestAccConnectionWithEnbledClients (13.82s)
=== RUN   TestAccConnectionSMS
--- PASS: TestAccConnectionSMS (7.09s)
=== RUN   TestAccConnectionEmail
--- PASS: TestAccConnectionEmail (1.63s)
=== RUN   TestAccConnectionSalesforce
--- PASS: TestAccConnectionSalesforce (7.10s)
=== RUN   TestAccConnectionGoogleOAuth2
--- PASS: TestAccConnectionGoogleOAuth2 (0.95s)
=== RUN   TestAccConnectionFacebook
--- PASS: TestAccConnectionFacebook (7.76s)
=== RUN   TestAccConnectionApple
--- PASS: TestAccConnectionApple (1.64s)
=== RUN   TestAccConnectionLinkedin
--- PASS: TestAccConnectionLinkedin (7.75s)
=== RUN   TestAccConnectionGitHub
--- PASS: TestAccConnectionGitHub (0.91s)
=== RUN   TestAccConnectionConfiguration
--- PASS: TestAccConnectionConfiguration (7.66s)
=== RUN   TestAccConnectionSAML
--- PASS: TestAccConnectionSAML (1.68s)
PASS
coverage: 26.9% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0	70.348s	coverage: 26.9% of statements
?   	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/debug	[no test files]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/random	0.254s	coverage: 0.0% of statements [no tests to run]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/validation	0.081s	coverage: 0.0% of statements [no tests to run]
?   	github.com/terraform-providers/terraform-provider-auth0/version	[no test files]

fbarbare commented:
Any updates on that PR? We need to configure those options too
@alexkappa

alexkappa commented:
Thank you @etwillbefine, I've merged your changes in. Will cut a new release soon with the validation fix in place, cheers!
